festim/hydrogen_transport_problem.py

```diff
@@ -480,13 +486,32 @@ def create_dirichletbc_form(self, bc):
 
         return form
 
+    def create_source_values_fenics(self):
+        """For each source create the value_fenics"""
+        for source in self.sources:
+            # create value_fenics for all F.Source objects
+            if isinstance(source, F.Source):
+                function_space_value = None
+                if callable(source.value):
+                    # if bc.value is a callable then need to provide a functionspace
+                    if not self.multispecies:
+                        function_space_value = source.species.sub_function_space
+                    else:
+                        function_space_value = source.species.collapsed_function_space
```

Comment:
Since the species is an attribute of the source, could this go inside `source.create_value_fenics()` ?

Reply:
Yeah this is true

Reply:
Actually no, we would still need to test if the model is multispecies or not. Same as for Dirichlet BC